src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddIndex(index);
+
+        if (index == size) {
+            add(value);
+        } else {
+            MyLinkedList.Node<T> succ = node(index);
+            final MyLinkedList.Node<T> pred = succ.prev;
+            final MyLinkedList.Node<T> newNode = new MyLinkedList.Node<>(pred, value, succ);
+            succ.prev = newNode;
+            if (pred == null) {
+                first = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        MyLinkedList.Node<T> x = node(index);
+        T oldVal = x.item;
+        x.item = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
-    public int size() {
-        return 0;
+    public boolean isEmpty() {
+        return size == 0;
     }
 
     @Override
-    public boolean isEmpty() {
-        return false;
+    public int size() {
+        return size;
+    }
+
+    private void checkIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException("There is no " + index + " index in LinkedList");
+        }
+    }
+
+    private void checkAddIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("There is no " + index + " index in LinkedList");
+        }
+    }
+
+    MyLinkedList.Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            MyLinkedList.Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            MyLinkedList.Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    private static class Node<E> {
```

Comment:
Don't use getters and setters in class Node. We have access to private fields of the inner class, so using getters or setters is redundant.